pkg/kubelet/kubelet_pods.go

```diff
@@ -1480,112 +1451,142 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
-				// they be removed at any time.
-				status = &oldStatus
-			} else {
-				// Apply some values from the old statuses as the default values.
-				status.RestartCount = oldStatus.RestartCount
-				status.LastTerminationState = oldStatus.LastTerminationState
+
+		for _, container := range containers {
+			status := &v1.ContainerStatus{
+				Name:  container.Name,
+				Image: container.Image,
+				State: defaultWaitingState,
 			}
+			oldStatus, found := oldStatuses[container.Name]
+			if found {
+				if oldStatus.State.Terminated != nil {
+					// Do not update status on terminated init containers as
+					// they be removed at any time.
+					status = &oldStatus
+				} else {
+					// Apply some values from the old statuses as the default values.
+					status.RestartCount = oldStatus.RestartCount
+					status.LastTerminationState = oldStatus.LastTerminationState
+				}
+			}
+			statuses[container.Name] = status
 		}
-		statuses[container.Name] = status
-	}
 
-	// Make the latest container status comes first.
-	sort.Sort(sort.Reverse(kubecontainer.SortContainerStatusesByCreationTime(podStatus.ContainerStatuses)))
-	// Set container statuses according to the statuses seen in pod status
-	containerSeen := map[string]int{}
-	for _, cStatus := range podStatus.ContainerStatuses {
-		cName := cStatus.Name
-		if _, ok := statuses[cName]; !ok {
-			// This would also ignore the infra container.
-			continue
-		}
-		if containerSeen[cName] >= 2 {
-			continue
-		}
-		status := convertContainerStatus(cStatus)
-		if containerSeen[cName] == 0 {
-			statuses[cName] = status
-		} else {
-			statuses[cName].LastTerminationState = status.State
+		// Make the latest container status comes first.
+		sort.Sort(sort.Reverse(kubecontainer.SortContainerStatusesByCreationTime(podStatus.ContainerStatuses)))
+		// Set container statuses according to the statuses seen in pod status
+		containerSeen := map[string]int{}
+		for _, cStatus := range podStatus.ContainerStatuses {
+			cName := cStatus.Name
+			if _, ok := statuses[cName]; !ok {
+				// This would also ignore the infra container.
+				continue
+			}
+			if containerSeen[cName] >= 2 {
+				continue
+			}
+			status := convertContainerStatus(cStatus)
+			if containerSeen[cName] == 0 {
+				statuses[cName] = status
+			} else {
+				statuses[cName].LastTerminationState = status.State
+			}
+			containerSeen[cName] = containerSeen[cName] + 1
 		}
-		containerSeen[cName] = containerSeen[cName] + 1
-	}
 
-	// Handle the containers failed to be started, which should be in Waiting state.
-	for _, container := range containers {
-		if isInitContainer {
-			// If the init container is terminated with exit code 0, it won't be restarted.
-			// TODO(random-liu): Handle this in a cleaner way.
-			s := podStatus.FindContainerStatusByName(container.Name)
-			if s != nil && s.State == kubecontainer.ContainerStateExited && s.ExitCode == 0 {
+		// Handle the containers failed to be started, which should be in Waiting state.
```

Comment:
Would it be clearer to put this status logic into a separate, testable function?